addons/udes_stock_cron/models/stock_picking.py

```diff
@@ -0,0 +1,201 @@
+        """
+
+        Picking = self.env["stock.picking"]
+        PickingType = self.env["stock.picking.type"]
+
+        if self:
+            picking_types = self.mapped("picking_type_id")
+        else:
+            picking_types = PickingType.search(
+                [("active", "=", True), ("u_num_reservable_pickings", "!=", 0)]
+            )
+
+        # We will either reserve up to the reservation limit or until all
+        # available picks have been reserved, depending on the value of
+        # u_num_reservable_pickings.
+        # However we must also take into account the atomic batch reservation
+        # flag (u_reserve_batches) and the handle partial flag
+        # (u_handle_partials).
+
+        for picking_type in picking_types:
+            _logger.info("Reserving stock for picking type %r.", picking_type)
+
+            # We want to reserve batches atomically, that is we will
+            # reserve pickings until all pickings in a batch have been
+            # assigned, even if we exceed the number of reservable pickings.
+            # However, the value of the handle partial flag is false we
+            # should not reserve stock if the batch cannot be completely
+            # reserved.
+            to_reserve = picking_type.u_num_reservable_pickings
+            reserve_all = to_reserve == -1
+            base_domain = [
+                ("picking_type_id", "=", picking_type.id),
+                ("state", "=", "confirmed"),
+            ]
+            limit = 1
+            processed = Picking.browse()
+
+            while reserve_all or to_reserve > 0:
+
+                if self:
+                    pickings = self.filtered(
+                        lambda p: p.picking_type_id == picking_type
+                    )
+                    # Removed processed pickings from self
+                    pickings -= processed
+                else:
+                    domain = base_domain[:]
+                    if processed:
+                        domain.append(("id", "not in", processed.ids))
+                    pickings = Picking.search(domain, limit=limit)
+
+                if not pickings:
+                    # No pickings left to process.
+                    # If u_num_reservable_pickings is -1, or there are
+                    # fewer available pickings that the limit, the loop must
+                    # terminate here.
+                    break
+
+                batch = pickings.mapped("batch_id")
+                if batch and batch.state == "draft":
+                    # Add to seen pickings so that we don't try to process
+                    # this batch again.
+                    processed |= batch.picking_ids
+                    continue
+
+                if batch and picking_type.u_reserve_batches:
+                    pickings = batch.picking_ids
+
+                # MPS: mimic Odoo's retry behaviour
+                tries = 0
+                while True:
+
+                    try:
```

Comment:
This function is fairly long -can we split up anywhere in a sensible place?

Reply:
I think this definitely needs to happen, but maybe this could be part of another story? I think migrating over the function and ensuring it works was the purpose of this story. Then any design changes could be made in another story.